### ISSUE TITLE
Support polymorphic models when fetching parent type from an API

### ISSUE THREAD

**Issue by ryandoherty** [emberjs/data#7357]
tldr; This block: https://github.com/emberjs/data/blob/ff4f9111fcfa7dd9e39804ed17f5af27a4a01378/packages/store/addon/-private/identifiers/cache.ts#L518 which verifies that the returned type matches the requested type is a problem for APIs that have polymorphic types. When we fetch data from the parent type's API endpoint, ember-data throws an exception because the api endpoint returns multiple types. But we do want all the subtypes of the parent type. 

### Example:
Given the JSON api:

```
/api/animals
/api/dogs
/api/cats
```

In our backend, 'dogs' and 'cats' are a subclass of 'animal', and queries to the /api/animals endpoint returns instances of cats and dogs. (This is via flask-restless & SQLAlchemy using the JSONAPI spec). Our Adapter is DS.JSONAPIAdapter

/api/animals returns:
```
{
  "data": [
    {
      "attributes": {
        "animal_type": "cat", <-- type column that SQLAlchemy uses to track the type of object to instantiate from the db row
        "created_at": "2020-10-19T13:29:07",
        "description": "my cat",
        "last_modified": "2020-10-21T15:04:08",
        "name": "foo"
      },
      "id": "1"
      },
      "type": "cat"
    },
    {
      "attributes": {
        "animal_type": "dog",
        "created_at": "2020-10-19T13:29:07",
        "description": "my dog",
        "last_modified": "2020-10-21T15:04:08",
        "name": "bar"
      },
      "id": "2"
      },
      "type": "dog"
    },
],
  "jsonapi": {
    "version": "1.0"
  },
  "links": {
    "first": "http://localhost:5000/api/animals?page[size]=10&page[number]=1",
    "last": "http://localhost:5000/api/animals?page[size]=10&page[number]=1",
    "next": null,
    "prev": null,
    "self": "/api/animals"
  },
  "meta": {
    "total": 2
  }
}
```

We have setup our ember models to have a parent type of 'Animal', 'Dog' and 'Cat' extend it. When we query the store for all animals like so: `this.store.fetchAll('animals')`, the reference error above is thrown because ember data expects only the type 'animal', but received 'dog' and 'cat'. 

What we'd like ember to do is handle this and return an array (or individual object) of the types received. There are many use cases where we want to query and display all animals in the database, but we have to query each subtype separately. We've done some hacks to our API and ember app to support this (forcing all types for /api/animals to be 'animal' and jumping through some hoops after loaded to recreate the actual type on the frontend)

There are also situations when we know the id of the animal, but not type and would like to use `this.store.findRecord('animal', id)` and get back an instance of whatever subtype it is.

I hope this is clear. From the comment in ember-data's code where the exception is thrown, it appears there is awareness that it would help to remove the check. Thank you!

**Comment by snewcomer** [comment]
@ryandoherty That block of code seems like a candidate for a PR!  Would you be willing to ✂️  and make a PR so we can discuss in our next meeting?  (we can also see if any tests fail).

**Comment by ryandoherty** [comment]
Will do thanks!

**Comment by ryandoherty** [comment]
Done! https://github.com/emberjs/data/pull/7421

**Comment by runspired** [comment]
For this situation you should use `query` (since as noted `findAll` is basically a query that then returns `peekAll` of the main type, but for polymorphism the main type may not include all of the records).

Alternatively you can write an array merged that watches the live arrays for all of the subclasses. Something like this:

```js
import Service, { inject } from '@ember/service';
import { tracked, cached } from '@glimmer/tracking';
import { A } from '@ember/array';
import { assert } from '@ember/debug';

const TypeMap = new Map([
  'animal', ['dog', 'cat']
]);

class LiveArray {
  @tracked storeArrays = A();
  
  @cached
  get data() {
    let a = [];
    this.storeArrays.forEach(arr => {
      a = a.concat(arr.toArray());
    });
    return a;
  }
}

export default class PolymorphicPeek extends Service {
  @inject store;
  
  liveArrays = new Map();
  
  peekAll(baseType) {
    assert(`${baseType} is not a polymorphic type. Use store.peekAll instead of polymorphicPeek.peekAll`, TypeMap.has(baseType);
    let liveArray = this.liveArrays.get(baseType);
    if (!liveArray) {
      const { store } = this;
      liveArray = new LiveArray();
      let types = TypeMap.get(baseType);
      liveArray.pushObject(store.peekAll(baseType));
      types.forEach(t => liveArray.pushObject(store.peekAll(t)));
      this.liveArrays.set(baseType, liveArray);
    }
    return liveArray;
  }
}

**Comment by runspired** [comment]
Hi @ryandoherty just wanted to check in and see if you needed anything more here. I am actively working on better polymorphism for relationships and so there may be an opportunity to improve the polymorphic-find ergonomics as well.

I suspect we can close this particular ticket since the underlying issue here was using `findAll` vs `query`, provided that `findRecord('animal', '1')` returning `'pet:1'` works correctly. [Since we have tests for that case](https://github.com/emberjs/data/blob/master/packages/-ember-data/tests/integration/identifiers/polymorphic-scenarios-test.ts), I'm going to close but happy to reopen if needed.
